pkg/ha/ha_parser_test.go

```diff
@@ -0,0 +1,343 @@
+		},
+		{
+			name:   "Test: HA enabled parse samples from standby prom instance.",
+			fields: fields{service: mockService},
+			args: args{tts: []prompb.TimeSeries{
+				{
+					Labels: []prompb.Label{
+						{Name: model.MetricNameLabelName, Value: "test"},
+						{Name: model.ReplicaNameLabel, Value: "replica2"},
+						{Name: model.ClusterNameLabel, Value: "cluster2"},
+					},
+					Samples: []prompb.Sample{
+						{Timestamp: sampleTimestamp, Value: 0.1},
+					},
+				},
+			}},
+			wantErr: false,
+			want:    nil,
+			want1:   0,
+		},
+		{
+			name:   "Test: HA enabled parse samples from leader prom instance.",
+			fields: fields{service: mockService},
+			args: args{tts: []prompb.TimeSeries{
+				{
+					Labels: []prompb.Label{
+						{Name: model.MetricNameLabelName, Value: "test"},
+						{Name: model.ReplicaNameLabel, Value: "replica1"},
+						{Name: model.ClusterNameLabel, Value: "cluster3"},
+					},
+					Samples: []prompb.Sample{
+						{Timestamp: sampleTimestamp, Value: 0.1},
+					},
+				},
+			}},
+			wantErr: false,
+			want: map[string][]model.SamplesInfo{
+				"test": {
+					{
+						Labels: &model.Labels{
+							Names:      []string{model.ClusterNameLabel, model.MetricNameLabelName, model.ReplicaNameLabel},
+							Values:     []string{"cluster3", "test", "replica1"},
+							MetricName: "test",
+							Str:        "\v\u0000__cluster__\b\u0000cluster3\b\u0000__name__\u0004\u0000test\v\u0000__replica__\b\u0000replica1",
+						},
+						SeriesID: -1,
+						Samples: []prompb.Sample{
+							{
+								Value:     0.1,
+								Timestamp: sampleTimestamp,
+							},
+						},
+					},
+				},
+			},
+			want1: 1,
+		},
+		{
+			name:   "Test: HA enabled parse samples from leader prom instance & samples are out of lease interval 1.",
+			fields: fields{service: mockService},
+			args: args{tts: []prompb.TimeSeries{
+				{
+					Labels: []prompb.Label{
+						{Name: model.MetricNameLabelName, Value: "test"},
+						{Name: model.ReplicaNameLabel, Value: "replica1"},
+						{Name: model.ClusterNameLabel, Value: "cluster3"},
+					},
+					Samples: []prompb.Sample{
+						{Timestamp: behindLeaseTimestamp, Value: 0.1},
+					},
+				},
+			}},
+			wantErr: false,
+			want:    nil,
+			want1:   0,
+		},
+		{
+			name:   "Test: HA enabled parse samples from leader prom instance & samples are out of lease interval 2.",
+			fields: fields{service: mockService},
+			args: args{tts: []prompb.TimeSeries{
+				{
+					Labels: []prompb.Label{
+						{Name: model.MetricNameLabelName, Value: "test"},
+						{Name: model.ReplicaNameLabel, Value: "replica1"},
+						{Name: model.ClusterNameLabel, Value: "cluster3"},
+					},
+					Samples: []prompb.Sample{
+						{Timestamp: aheadLeaseTimestamp, Value: 0.1},
+					},
+				},
+			}},
+			wantErr: false,
+			want:    nil,
+			want1:   0,
+		},
+		{
+			name:   "Test: HA enabled parse samples from standby prom instance. readLockState returns the updated leader as standby prom instance.",
```

Comment:
@blagojts THis testcase is failing as this condition `currentMaxT.After(h.maxTimeSeen)` is failing